### PR TITLE
Switch to using winapi for file I/O

### PR DESCRIPTION
**Pull request by thrimbor**
Due to requiring properly mounted devices, this will be a breaking change for some users, so I recommend to merge this after #14.

This changes the file I/O functions in the xbox backend to use the winapi functions instead of the old X*-functions. I used PDCLib's shepherd-branch for flag reference, the `_PDCLIB_w32errno()` function for translating winapi errors to errno values also comes from there.

I gave this some testing, and it seemed to work fine, but I do recommend everyone to test their code with this to make sure this doesn't break in some specific cases.

### CODE REVIEW

File: platform/xbox/functions/_PDCLIB/_PDCLIB_w32errno.c
Comment (JayFoxRox) [inline platform/xbox/functions/_PDCLIB/_PDCLIB_w32errno.c]: What does this mean? Should it be important to us?

File: platform/xbox/functions/_PDCLIB/_PDCLIB_rename.c
Comment (JayFoxRox) [inline platform/xbox/functions/_PDCLIB/_PDCLIB_rename.c]: `MoveFileA`

File: platform/xbox/functions/_PDCLIB/_PDCLIB_close.c
Comment (JayFoxRox) [inline platform/xbox/functions/_PDCLIB/_PDCLIB_close.c:29]: A bit surprising that this isn't setting `errno` in some way

File: platform/xbox/include/pdclib/_PDCLIB_config.h
Comment (JayFoxRox) [inline platform/xbox/include/pdclib/_PDCLIB_config.h:283]: nit: Why isn't this `HANDLE`?
I assume an include would drag too much stuff into this file?

Still feels a bit dirty

File: platform/xbox/functions/_PDCLIB/_PDCLIB_w32errno.c
Comment (thrimbor) [reply platform/xbox/functions/_PDCLIB/_PDCLIB_w32errno.c]: No, that's only relevant for Watcom compiler users. That comment is only there because I copied the file and left it untouched.

File: platform/xbox/functions/_PDCLIB/_PDCLIB_rename.c
Comment (thrimbor) [reply platform/xbox/functions/_PDCLIB/_PDCLIB_rename.c]: Changed.

File: platform/xbox/functions/_PDCLIB/_PDCLIB_close.c
Comment (thrimbor) [reply platform/xbox/functions/_PDCLIB/_PDCLIB_close.c:29]: That was an oversight, it does set it now.

File: platform/xbox/include/pdclib/_PDCLIB_config.h
Comment (thrimbor) [reply platform/xbox/include/pdclib/_PDCLIB_config.h:283]: > nit: Why isn't this `HANDLE`?
> I assume an include would drag too much stuff into this file?

Exactly, dragging in windows headers with every C standard library header doesn't seem like a good idea to me.

Comment (JayFoxRox) [reply platform/xbox/include/pdclib/_PDCLIB_config.h:283]: I had originally resolved this discussion; but while doing another review pass I noticed that we should at least describe this in a comment (= that it is a `HANDLE`, and that it has to match the typedef from that header)

File: platform/xbox/functions/_PDCLIB/_PDCLIB_w32errno.c
Comment (JayFoxRox) [inline platform/xbox/functions/_PDCLIB/_PDCLIB_w32errno.c]: This prototype seems fairly pointless?

**Comment by JayFoxRox** [comment]
I still want to test this; please wait with merge.

File: platform/xbox/functions/_PDCLIB/_PDCLIB_close.c
Comment (JayFoxRox) [inline platform/xbox/functions/_PDCLIB/_PDCLIB_close.c]:
```
samples/fileio-test/../../lib/pdclib/platform/xbox/functions/_PDCLIB/_PDCLIB_close.c:25:9: warning: implicit declaration of function '_PDCLIB_w32errno' is invalid in C99 [-Wimplicit-function-declaration]
        _PDCLIB_w32errno();
        ^
```

(To even reach this point, I had to rebase to the latest master, as I otherwise run into issues with lack of #14 ?)

File: platform/xbox/functions/_PDCLIB/_PDCLIB_w32errno.c
Comment (JayFoxRox) [reply platform/xbox/functions/_PDCLIB/_PDCLIB_w32errno.c]: Reminder @thrimbor

File: platform/xbox/include/pdclib/_PDCLIB_config.h
Comment (JayFoxRox) [reply platform/xbox/include/pdclib/_PDCLIB_config.h:283]: Reminder @thrimbor

**Comment by JayFoxRox** [comment]
I'm not sure if I did something wrong, but the test results were *bad*. I'm not expecting all of these to work / be changed, but the results are unexpected:

- **`assert(fopen("c:\\filec.test", "wb") != NULL);`**
- **`assert(fopen("d:\\filed.test", "wb") != NULL);`**

This works, as expected.

- **`//assert(fopen("C:\\fileC.test", "wb") != NULL); // Broken`**
- **`//assert(fopen("D:\\fileD.test", "wb") != NULL); // Broken`**

This does not work, and that really surprised me. If anything, I'd expect this to work, but not the lowercase variant. I mounted my filesystem with `nxMountDrive('C', "\\Device\\Harddisk0\\Partition2\\");` (uppercase `C`)

Also [see this MSDN doc](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file):

> Volume designators (drive letters) are similarly case-insensitive. For example, "D:\" and "d:\" refer to the same volume.

*(I did not check if this might have to do with the lowercase variants already existing)*

- **`//assert(fopen("c:/filecf.test", "wb") != NULL); // Broken`**
- **`//assert(fopen("C:/fileCf.test", "wb") != NULL); // Broken`**
- **`//assert(fopen("d:/filedf.test", "wb") != NULL); // Broken`**
- **`//assert(fopen("D:/fileDf.test", "wb") != NULL); // Broken`**

These were unexpected, because [the MSDN even says](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilea):

> The name of the file or device to be created or opened. You may use either forward slashes (/) or backslashes () in this name.

I did not check if this is also an issue with `CreateFile` and other winapi functions, but I'd assume so (hence the quote from MSDN).

If the kernel doesn't handle this, then we'll probably need a translation function in our winapi.

- **`//assert(fopen(".\\filedot.test", "wb") != NULL); // Broken`**
- **`//assert(fopen("./filedotf.test", "wb") != NULL); // Broken`**
- **`//assert(fopen("file.test", "wb") != NULL); // Broken`**

These do not work. This was somewhat expected (I think). However, this is also severly limiting. A lot of the software I've seen uses such filepaths.
I don't have a Windows machine to test these, but I'd assume their `CreateFile` to handle this, or at the very least I'd assume that their libc handles this for portability.

*(I'm aware that `D:\\` was advocated for Xbox, but I'm not sure if we should really expect people to keep using that; I'm also not sure how the XDK handled relative paths)*

- **Things I did not try**

I did not try parent-paths such as `..` or path like `\\foo` to be in the root of the current drive (I'd assume the MS libc to allow these things).

File: platform/xbox/include/pdclib/_PDCLIB_config.h
Comment (thrimbor) [reply platform/xbox/include/pdclib/_PDCLIB_config.h:283]: I added a comment.

File: platform/xbox/functions/_PDCLIB/_PDCLIB_w32errno.c
Comment (thrimbor) [reply platform/xbox/functions/_PDCLIB/_PDCLIB_w32errno.c]: Probably, but as I said above, I didn't touch the file contents.

File: platform/xbox/functions/_PDCLIB/_PDCLIB_close.c
Comment (thrimbor) [reply platform/xbox/functions/_PDCLIB/_PDCLIB_close.c]: The rebase probably was necessary because we had nxdk changes in the meantime. I rebased the branch to fix it, and added the missing declaration.

**Comment by thrimbor** [comment]
> I'm not sure if I did something wrong, but the test results were _bad_. I'm not expecting all of these to work / be changed, but the results are unexpected:
> 
>     * **`assert(fopen("c:\\filec.test", "wb") != NULL);`**
> 
>     * **`assert(fopen("d:\\filed.test", "wb") != NULL);`**
> 
> This works, as expected.
> 
>     * **`//assert(fopen("C:\\fileC.test", "wb") != NULL); // Broken`**
>     * **`//assert(fopen("D:\\fileD.test", "wb") != NULL); // Broken`**

Not sure what's happening here - did you close the file streams / reboot before? We don't open with `FILE_SHARE_WRITE`, so there can only be one file stream with write access to the file.

>     * **`//assert(fopen("c:/filecf.test", "wb") != NULL); // Broken`**
>     * **`//assert(fopen("C:/fileCf.test", "wb") != NULL); // Broken`**
>     * **`//assert(fopen("d:/filedf.test", "wb") != NULL); // Broken`**
>     * **`//assert(fopen("D:/fileDf.test", "wb") != NULL); // Broken`**
> These were unexpected, because [the MSDN even says](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilea):
> 
> > The name of the file or device to be created or opened. You may use either forward slashes (/) or backslashes () in this name.
> 
> I did not check if this is also an issue with `CreateFile` and other winapi functions, but I'd assume so (hence the quote from MSDN).
> 
> If the kernel doesn't handle this, then we'll probably need a translation function in our winapi.

Yeah that's a shortcoming of our current implementation in winapi. `CreateFileA` should handle those slashes.

> 
>     * **`//assert(fopen(".\\filedot.test", "wb") != NULL); // Broken`**
>     * **`//assert(fopen("./filedotf.test", "wb") != NULL); // Broken`**
>     * **`//assert(fopen("file.test", "wb") != NULL); // Broken`**
> 
> These do not work. This was somewhat expected (I think). However, this is also severly limiting. A lot of the software I've seen uses such filepaths.
> I don't have a Windows machine to test these, but I'd assume their `CreateFile` to handle this, or at the very least I'd assume that their libc handles this for portability.
> 
> _(I'm aware that `D:\\` was advocated for Xbox, but I'm not sure if we should really expect people to keep using that; I'm also not sure how the XDK handled relative paths)_

Handling working directories seems to usually involve quite a bit of kernel assistance, which is not provided on the Xbox (that's why the OpenXDK code had such a hack for it). The XDK solution (I'm deducing this from publicly available source such as XBMC) seems to be to just not support it at all, and use absolute paths everywhere.

I thought I could get by without touching the path strings in the winapi functions, but if I want to handle forward slashes I'll have to do that anyway (XDK probably doesn't support them? idk). I'll have to think about that for while to figure out how I would implement all this, and how it impacts stuff like `MoveFileA`.

**Comment by JayFoxRox** [comment]
> Not sure what's happening here - did you close the file streams / reboot before? We don't open with FILE_SHARE_WRITE, so there can only be one file stream with write access to the file.

That does make a lot of sense. I forgot about this.

> [...] touching the path strings in the winapi functions [...].
> I'll have to think about that for while to figure out how I would implement all this [...].

I'm fine with this not being implemented in this PR, but we'll need issues for things that are expected to work, and probably also a nxdk wiki article or some other form of documentation which documents what works, what doesn't work (but should work = links to issues) and non-goals.

Because clearly a lot of our winapi will diverge from what the MSDN docments in unexpected ways, and win32 libc code will not be portable.

> [...] but I do recommend everyone to test their code with this to make sure this doesn't break in some specific cases.

I'm sure that many applications will break.
Given my results I skimmed over some nxdk code from https://github.com/XboxDev/nxdk/wiki/Projects-using-nxdk we can find a couple of spots in different projects very easily (I did not test these):

- https://github.com/Ryzee119/OpenXenium/blob/1e11059c5a759365e45ffc97ae1550b9c695678d/Xenium-Tools/main.c#L89
- https://github.com/Cxbx-Reloaded/xbox_kernel_test_suite/blob/c3dc177e47c4d1ffbe706ebb07a12d952164be3d/main.c#L110
- https://github.com/dracc/NevolutionX/blob/2aeb9ce77f68a3183f269bef196b05a309b0f88c/Includes/font.cpp#L9 via https://github.com/dracc/NevolutionX/blob/2aeb9ce77f68a3183f269bef196b05a309b0f88c/main.cpp#L68
- https://github.com/XboxDev/nxdk/blob/a944b38b1bdd1753dc484305e061f5eb98a05c74/samples/sdl_ttf/main.c#L55
- I'm also sure ports like DOOOM are affected, but I had trouble pinpointing a specific instance.

I'm not sure how to deal with this, because clearly this repository isn't gaining enough attention for this breaking-change (nobody had complained since November, despite obvious issues with backwards compatibility).

**Comment by JayFoxRox** [comment]
I think we should merge this as-is and iterate later.

However, first, we should create issues about this and add some documentation somewhere (nxdk wiki?), which lists what we do / do not support and what we plan to support / do not plan to support (for winapi and C stdio), including best practices how to avoid this. We'll probably also have to adapt some samples when updating the submodule.

If this breaks apps, then people should be able to look at the docs to decide a plan of action, or they can stick to an older version of nxdk for now.
But clearly there is interest in advancing nxdk, and this PR is a step in that direction (even if it means dropping some defacto standard C features).

File: platform/xbox/functions/_PDCLIB/_PDCLIB_fillbuffer.c
Comment (JayFoxRox) [inline platform/xbox/functions/_PDCLIB/_PDCLIB_fillbuffer.c]: @sam-itt Is this logic enough? (Missing `_PDCLIB_EOFFLAG` handling?)

https://github.com/XboxDev/nxdk-pdclib/pull/22/files#diff-51228aebb715e9137534cdecdb53aa0dR53 does a bit more and you said `feof` was broken (or is this fixed independently in this PR already?).

Are there any other changes in your PR that are different than what is done here? Could you please leave review comments asap?

Comment (sam-itt) [reply platform/xbox/functions/_PDCLIB/_PDCLIB_fillbuffer.c]: feof is currently broken because XReadFile doesn't sets the flag and doesn't give enough info to do it at the _PDCLIB_fillbuffer level. 

You *need* to detect EOF and set stream->status  accordingly here. Otherwise it won't be set and thus feof won't work.

**Review by sam-itt — CHANGES_REQUESTED** [review]
I suggest using error handling code from #22 it has more cases, is more generic and uses PDCLIB internal error codes like other functions in the lib do.

**Review by sam-itt — CHANGES_REQUESTED** [review]
Merging this will break relative path without notice. I suggest adding a #warning in stdio.h to let users know this at compile time.

File: platform/xbox/functions/_PDCLIB/_PDCLIB_w32errno.c
Comment (sam-itt) [inline platform/xbox/functions/_PDCLIB/_PDCLIB_w32errno.c]: It seems that other part of the lib don't assign errno directly but rather uses _PDCLIB_errno which in fact is a function call that returns the address of errno. Thread safety maybe ?

File: platform/xbox/functions/_PDCLIB/_PDCLIB_fillbuffer.c
Comment (thrimbor) [reply platform/xbox/functions/_PDCLIB/_PDCLIB_fillbuffer.c]: EOF is already handled: https://github.com/thrimbor/nxdk-pdclib/blob/fcd7865af33af10c3b5ab31139d79d5b2fee8282/platform/xbox/functions/_PDCLIB/_PDCLIB_fillbuffer.c#L28-L33

The additional EOF-handling code in #22 is never actually used, because `ReadFile` doesn't report EOF this way - it only does that when using overlapped IO, which we don't do here (nxdk doesn't even support that yet). When using synchronous IO, `ReadFile` reports EOF by returning `TRUE` while setting `lpNumberOfBytesRead` to zero.

If you have any issues regarding EOF handling, please provide a minimal test case.

File: platform/xbox/functions/_PDCLIB/_PDCLIB_w32errno.c
Comment (thrimbor) [reply platform/xbox/functions/_PDCLIB/_PDCLIB_w32errno.c]: `errno` is the same as `*_PDCLIB_errno()`.

**Comment by thrimbor** [comment]
> I suggest using error handling code from #22 it has more cases, is more generic and uses PDCLIB internal error codes like other functions in the lib do.

I'm not opposed to that idea, but I'd like to know where that mapping comes from first, and I'm also not a fan of the `fallback_errno` parameter.

**Comment by thrimbor** [comment]
> Merging this will break relative path without notice. I suggest adding a #warning in stdio.h to let users know this at compile time.

I don't think that's a good solution. It would introduce warnings where nothing's actually wrong, and it also may bury those warnings, e.g. when users are using SDL to open a file, they won't get the warning where the actual problem is.

**Comment by sam-itt** [comment]
> I'm not opposed to that idea, but I'd like to know where that mapping comes from first
Part from cygnus, part from myself.
>, and I'm also not a fan of the `fallback_errno` parameter.
You'll always have a fallback. At least here it's explicit.

**Comment by sam-itt** [comment]
>. when users are using SDL to open a file, they won't get the warning where the actual problem is.
Good point. So it's either
- a) merge as-is and fix code that breaks (i.e) relative path
-b) Add some code in fopen or CreateFileA to handle relative paths before merging.

File: platform/xbox/functions/_PDCLIB/_PDCLIB_fillbuffer.c
Comment (sam-itt) [reply platform/xbox/functions/_PDCLIB/_PDCLIB_fillbuffer.c]: feof() doesn't work with the current X*File* based code. If it works with your ReadFile-based implementation that's allright by me.

**Comment by thrimbor** [comment]
I've updated the PR with the error translations from #22. I did some minor changes to it, like formatting changes and changing the fallback (even with error checking in `strerror` I'm not comfortable with it just returning an out of bounds value). I also ended up squashing the changes into my commit because trying to rebase the changes into a meaningful history was quite annoying - I hope you don't mind @sam-itt.

**Comment by sam-itt** [comment]
I don't mind at all: Glad it's merged!

**Comment by JayFoxRox** [comment]
Last call for vetos, tests or reviews. I'll do a final pass tonight and merge it then (if necessary, untested or without another extensive review pass).

This sort of patch has been bikeshedded for too long already.

File: platform/xbox/functions/stdio/remove.c
Comment (JayFoxRox) [inline platform/xbox/functions/stdio/remove.c]: The signature should probably be:  `int _PDCLIB_w32errno( DWORD werror );` (like elsewhere in your code) and this call should probably be:

```c
errno = _PDCLIB_w32errno(GetLastError());
```

File: platform/xbox/functions/_PDCLIB/_PDCLIB_open.c
Comment (JayFoxRox) [inline platform/xbox/functions/_PDCLIB/_PDCLIB_open.c]: ~~nit:~~ Why doesn't this return `_PDCLIB_fd_t`? What does upstream pdclib do here in its reference implementation?

File: platform/xbox/functions/_PDCLIB/_PDCLIB_fillbuffer.c
Comment (JayFoxRox) [inline platform/xbox/functions/_PDCLIB/_PDCLIB_fillbuffer.c:48]: nit: Why does your code use `errno`, but our old code had `_PDCLIB_errno`?

File: platform/xbox/functions/_PDCLIB/_PDCLIB_w32errno.c
Comment (JayFoxRox) [reply platform/xbox/functions/_PDCLIB/_PDCLIB_w32errno.c]: I also noticed this; in my opinion, we should stick to whatever, pdclib does / prefers.

- Will make it easier to compare to upstream and will also allow for possible changes.
- We also don't know if `_PDCLIB_errno` might be made per-thread (or something) in the future. Maybe it exists for good reasons that we don't understand.

It's a nit from me objectively (I don't think anything bad will happen). Subjectively I feel strongly (I dislike diverging from upstream).

Comment (thrimbor) [reply platform/xbox/functions/_PDCLIB/_PDCLIB_w32errno.c]: > We also don't know if _PDCLIB_errno might be made per-thread (or something) in the future. Maybe it exists for good reasons that we don't understand.

Yes, that's the case now in upstream PDCLib, and it's the exact reason why `errno` is the same as `*_PDCLIB_errno()`, to support this kind of change.
The example port in PDCLib seems to prefer using ``*_PDCLIB_errno()` directly, so I adjusted my code.

File: platform/xbox/functions/stdio/remove.c
Comment (thrimbor) [reply platform/xbox/functions/stdio/remove.c]: Oops, yes it should.

File: platform/xbox/functions/_PDCLIB/_PDCLIB_open.c
Comment (thrimbor) [reply platform/xbox/functions/_PDCLIB/_PDCLIB_open.c]: PDCLib's example port uses `int` here (which is the same as its `_PDCLIB_fd_t`.

Comment (JayFoxRox) [reply platform/xbox/functions/_PDCLIB/_PDCLIB_open.c]: CC @DevSolar. This feels like a bug.
I think we should replace it by `_PDCLIB_fd_t`?

Comment (DevSolar) [reply platform/xbox/functions/_PDCLIB/_PDCLIB_open.c]: It is indeed a bug, which got fixed in SVN @856 / GIT 7a65248 (2019-12-19). Yes, _PDCLIB_fd_t would be the correct thing to return here (as the return value would be inserted as such into struct _PDCLIB_file_t / FILE).

Comment (JayFoxRox) [reply platform/xbox/functions/_PDCLIB/_PDCLIB_open.c]: In that case it's no longer a nit for me; please fix @thrimbor

File: platform/xbox/functions/_PDCLIB/_PDCLIB_w32errno.c
Comment (DevSolar) [reply platform/xbox/functions/_PDCLIB/_PDCLIB_w32errno.c]: If I may shine some light on this... there's a section in P.J. Plaugher's book "The Standard C Library" on the merits of using "lazy evaluation" of errno when working with the FPU. Reading the FPU status flags, he claims, basically stalls the FPU pipeline, so it would be better to have the FPU carry on _until the errno value is actually requested_, instead of setting errno after each operation.

The other point was, indeed, thread safety -- I figured I might need some initialization code to run the first time a thread queried errno.

I haven't dabbled with the FPU yet, and the thread safety turned out to be much simpler (not requiring any dynamic setup), but I didn't know that at the time I implemented `<errno.h>` (and came up with `*_PDCLIB_errno_func()` as a convenient backdoor to allow errno-handling code).

As for what is "preferred", I haven't actually layed down a "rule" for that. In some places (platform code) I am handling **PDCLib** errno **and** native errno in the same source; here, of course I have to go through `*_PDCLIB_errno_func()` and `_PDCLIB_E...` while e.g. /usr/include/errno.h is included. I could use just plain `errno` in other source files, and am a bit torn between readability and consistency. I've shelved making any "clean-up sweeps" on errno right now because I've got so many construction sites already. But if it helps you with **your* work, I could give it a quick once-over.

File: platform/xbox/functions/_PDCLIB/_PDCLIB_open.c
Comment (thrimbor) [reply platform/xbox/functions/_PDCLIB/_PDCLIB_open.c]: Fixed.

File: platform/xbox/functions/_PDCLIB/_PDCLIB_w32errno.c
Comment (JayFoxRox) [reply platform/xbox/functions/_PDCLIB/_PDCLIB_w32errno.c]: This PR is consistent now. We can adapt whatever pdclib chooses when doing a rebase in the future.

**Comment by JayFoxRox** [comment]
I'll merge this now, despite this not being ideal:

- Ideally we'd have had another rebase of pdclib first (to get rid of bugs before doing a larger breaking change), but I think we gave people enough time to review their code.
- Ideally we'd have fixed gmtime first; I also planned to do this last week but didn't find the time.

We'll have to remember to fix samples in nxdk (like samples/sdl_ttf) when updating the submodule in nxdk (which should happen as soon as possible, so that we don't group too many changes together).

Merged.

**Pull request merged**
